In [34]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

# Any results you write to the current directory are saved as output.

In [35]:
data= pd.read_csv('train.csv')

In [36]:
data.head()

id  distance taxi_type  months_of_activity  customer_score  \
0  T0005891966     59.30         C                 1.0             NaN   
1  T0005902453     20.68         B                 0.0         2.79222   
2  T0005847289     64.83         B                 1.0         2.97482   
3  T0005741279     25.51         C                 0.0         2.90334   
4  T0005711058     20.16         D                 6.0         2.62301   

  customer_score_confidence drop_location_type  ratings_given_by_cust  \
0                       NaN                  B                1.18625   
1                         C                  B                1.91750   
2                         C                  A                3.62125   
3                         A                  B                4.76750   
4                         B                  D                2.03875   

   num_of_cancelled_trips  anon_var_1  anon_var_2  anon_var_3     sex  \
0                       0         NaN          50          71    Male   
1                       1        31.0          50          65    Male   
2                       0        65.0          49          84    Male   
3                       2        46.0          43          60  Female   
4                       0         NaN          52          71  Female   

   pricing_category  
0                 2  
1                 2  
2                 2  
3                 3  
4                 3

In [37]:
data.describe()

distance  months_of_activity  customer_score  \
count  78997.000000        75464.000000    66956.000000   
mean      44.201348            6.014179        2.801450   
std       25.544393            3.629048        0.226158   
min        1.530000            0.000000        1.656960   
25%       24.540000            3.000000        2.653680   
50%       38.200000            6.000000        2.796885   
75%       60.740000           10.000000        2.946198   
max      109.230000           10.000000        4.875110   

       ratings_given_by_cust  num_of_cancelled_trips    anon_var_1  \
count           78997.000000            78997.000000  36578.000000   
mean                2.853010                0.784042     64.249768   
std                 0.981382                1.036743     21.800992   
min                 0.001250                0.000000     30.000000   
25%                 2.153750                0.000000     46.000000   
50%                 2.900000                0.000000     61.000000   
75%                 3.587500                1.000000     80.000000   
max                 5.000000                8.000000    200.000000   

         anon_var_2    anon_var_3  pricing_category  
count  78997.000000  78997.000000      78997.000000  
mean      51.202919     75.102561          2.154474  
std        4.995485     11.595299          0.738508  
min       40.000000     52.000000          1.000000  
25%       48.000000     67.000000          2.000000  
50%       50.000000     74.000000          2.000000  
75%       54.000000     82.000000          3.000000  
max      124.000000    206.000000          3.000000

In [38]:
x= data.isna().sum() 
print(x)


id                               0
distance                         0
taxi_type                    12152
months_of_activity            3533
customer_score               12041
customer_score_confidence    12041
drop_location_type               0
ratings_given_by_cust            0
num_of_cancelled_trips           0
anon_var_1                   42419
anon_var_2                       0
anon_var_3                       0
sex                              0
pricing_category                 0
dtype: int64


In [39]:
x= data['anon_var_3']-data['anon_var_2']

x.describe()

count    78997.000000
mean        23.899642
std          8.951540
min         10.000000
25%         17.000000
50%         22.000000
75%         29.000000
max        158.000000
dtype: float64

In [40]:
#data['diff']= data['anon_var_3']-data['anon_var_2']

In [41]:
data.head()

id  distance taxi_type  months_of_activity  customer_score  \
0  T0005891966     59.30         C                 1.0             NaN   
1  T0005902453     20.68         B                 0.0         2.79222   
2  T0005847289     64.83         B                 1.0         2.97482   
3  T0005741279     25.51         C                 0.0         2.90334   
4  T0005711058     20.16         D                 6.0         2.62301   

  customer_score_confidence drop_location_type  ratings_given_by_cust  \
0                       NaN                  B                1.18625   
1                         C                  B                1.91750   
2                         C                  A                3.62125   
3                         A                  B                4.76750   
4                         B                  D                2.03875   

   num_of_cancelled_trips  anon_var_1  anon_var_2  anon_var_3     sex  \
0                       0         NaN          50          71    Male   
1                       1        31.0          50          65    Male   
2                       0        65.0          49          84    Male   
3                       2        46.0          43          60  Female   
4                       0         NaN          52          71  Female   

   pricing_category  
0                 2  
1                 2  
2                 2  
3                 3  
4                 3

# Handling missing values.

In [42]:
'''
#percentage of NaN per column
print("anon_var_1:", x['anon_var_1']/data.shape[0],"%")
print("customer score:", x['customer_score']/data.shape[0],"%")
print("customer_score_confidence:", x['customer_score_confidence']/data.shape[0],"%")
'''

'\n#percentage of NaN per column\nprint("anon_var_1:", x[\'anon_var_1\']/data.shape[0],"%")\nprint("customer score:", x[\'customer_score\']/data.shape[0],"%")\nprint("customer_score_confidence:", x[\'customer_score_confidence\']/data.shape[0],"%")\n'

There are a lot of NaN values in the dataset. "anon_var_1" has 42419 NaNs. Its better to ignore it.

In [43]:
# We don't need id as well
train= data.drop(['anon_var_1','id'],axis=1)

In [44]:
train.describe()

distance  months_of_activity  customer_score  \
count  78997.000000        75464.000000    66956.000000   
mean      44.201348            6.014179        2.801450   
std       25.544393            3.629048        0.226158   
min        1.530000            0.000000        1.656960   
25%       24.540000            3.000000        2.653680   
50%       38.200000            6.000000        2.796885   
75%       60.740000           10.000000        2.946198   
max      109.230000           10.000000        4.875110   

       ratings_given_by_cust  num_of_cancelled_trips    anon_var_2  \
count           78997.000000            78997.000000  78997.000000   
mean                2.853010                0.784042     51.202919   
std                 0.981382                1.036743      4.995485   
min                 0.001250                0.000000     40.000000   
25%                 2.153750                0.000000     48.000000   
50%                 2.900000                0.000000     50.000000   
75%                 3.587500                1.000000     54.000000   
max                 5.000000                8.000000    124.000000   

         anon_var_3  pricing_category  
count  78997.000000      78997.000000  
mean      75.102561          2.154474  
std       11.595299          0.738508  
min       52.000000          1.000000  
25%       67.000000          2.000000  
50%       74.000000          2.000000  
75%       82.000000          3.000000  
max      206.000000          3.000000

In [45]:
train.isna().sum() 

distance                         0
taxi_type                    12152
months_of_activity            3533
customer_score               12041
customer_score_confidence    12041
drop_location_type               0
ratings_given_by_cust            0
num_of_cancelled_trips           0
anon_var_2                       0
anon_var_3                       0
sex                              0
pricing_category                 0
dtype: int64

In [46]:
train.head(15)

train['drop_location_type'].unique()
train['taxi_type'].unique()

array(['C', 'B', 'D', 'A', 'E', nan], dtype=object)

In [47]:
# Lets see the taxi type count
from collections import Counter
taxi_type_counter= Counter(train['taxi_type'])
customer_score_confidence_counter= Counter(train['customer_score_confidence'])
drop_location_type_counter= Counter(train['drop_location_type'])
print(taxi_type_counter)
print(customer_score_confidence_counter)
print(drop_location_type_counter)

Counter({'B': 18644, 'C': 16827, 'A': 13011, nan: 12152, 'D': 11348, 'E': 7015})
Counter({'B': 24148, 'C': 21639, 'A': 21169, nan: 12041})
Counter({'A': 46616, 'B': 17676, 'C': 4534, 'D': 3970, 'E': 1616, 'F': 1145, 'G': 886, 'H': 767, 'I': 487, 'K': 416, 'J': 403, 'L': 371, 'N': 60, 'M': 50})


Taxi type B is the most used. So I think that is the average quality taxi followed by C. So we think that E is the most costly car.

In [48]:
train_numerical= train.replace({'taxi_type': {'A':1, 'B': 2,'C':3,'D':4,'E':5}})

In [49]:
#train_numerical.replace({'customer_score_confidence':{'A':1, 'B': 3,'C':2}}, inplace=True)

In [50]:
'''
train_numerical.replace({'drop_location_type':{'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E':5,
                                               'F': 6, 'G': 7, 'H': 8, 'I': 9, 'K': 10, 'J': 11, 
                                               'L': 12, 'N': 13, 'M': 14}},inplace=True)
'''
'''

train_numerical.replace({'drop_location_type':{'A': 11, 'B':13, 'C': 12, 'D':14, 'E':10,
                                               'F': 9, 'G': 8, 'H': 7, 'I': 6, 'K': 5, 'J': 4, 
                                               'L': 3, 'N': 2, 'M': 1}},inplace=True)
'''

"\n\ntrain_numerical.replace({'drop_location_type':{'A': 11, 'B':13, 'C': 12, 'D':14, 'E':10,\n                                               'F': 9, 'G': 8, 'H': 7, 'I': 6, 'K': 5, 'J': 4, \n                                               'L': 3, 'N': 2, 'M': 1}},inplace=True)\n"

In [51]:
train_numerical.head()

distance  taxi_type  months_of_activity  customer_score  \
0     59.30        3.0                 1.0             NaN   
1     20.68        2.0                 0.0         2.79222   
2     64.83        2.0                 1.0         2.97482   
3     25.51        3.0                 0.0         2.90334   
4     20.16        4.0                 6.0         2.62301   

  customer_score_confidence drop_location_type  ratings_given_by_cust  \
0                       NaN                  B                1.18625   
1                         C                  B                1.91750   
2                         C                  A                3.62125   
3                         A                  B                4.76750   
4                         B                  D                2.03875   

   num_of_cancelled_trips  anon_var_2  anon_var_3     sex  pricing_category  
0                       0          50          71    Male                 2  
1                       1          50          65    Male                 2  
2                       0          49          84    Male                 2  
3                       2          43          60  Female                 3  
4                       0          52          71  Female                 3

In [52]:
train_hot_encoded= pd.get_dummies(train_numerical)

In [53]:
train_hot_encoded.head()
#train_hot_encoded.shape

distance  taxi_type  months_of_activity  customer_score  \
0     59.30        3.0                 1.0             NaN   
1     20.68        2.0                 0.0         2.79222   
2     64.83        2.0                 1.0         2.97482   
3     25.51        3.0                 0.0         2.90334   
4     20.16        4.0                 6.0         2.62301   

   ratings_given_by_cust  num_of_cancelled_trips  anon_var_2  anon_var_3  \
0                1.18625                       0          50          71   
1                1.91750                       1          50          65   
2                3.62125                       0          49          84   
3                4.76750                       2          43          60   
4                2.03875                       0          52          71   

   pricing_category  customer_score_confidence_A    ...     \
0                 2                            0    ...      
1                 2                            0    ...      
2                 2                            0    ...      
3                 3                            1    ...      
4                 3                            0    ...      

   drop_location_type_G  drop_location_type_H  drop_location_type_I  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   drop_location_type_J  drop_location_type_K  drop_location_type_L  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   drop_location_type_M  drop_location_type_N  sex_Female  sex_Male  
0                     0                     0           0         1  
1                     0                     0           0         1  
2                     0                     0           0         1  
3                     0                     0           1         0  
4                     0                     0           1         0  

[5 rows x 28 columns]

In [54]:
train_hot_encoded1= train_hot_encoded.values

In [55]:
train= train_hot_encoded

In [56]:
train.head()

distance  taxi_type  months_of_activity  customer_score  \
0     59.30        3.0                 1.0             NaN   
1     20.68        2.0                 0.0         2.79222   
2     64.83        2.0                 1.0         2.97482   
3     25.51        3.0                 0.0         2.90334   
4     20.16        4.0                 6.0         2.62301   

   ratings_given_by_cust  num_of_cancelled_trips  anon_var_2  anon_var_3  \
0                1.18625                       0          50          71   
1                1.91750                       1          50          65   
2                3.62125                       0          49          84   
3                4.76750                       2          43          60   
4                2.03875                       0          52          71   

   pricing_category  customer_score_confidence_A    ...     \
0                 2                            0    ...      
1                 2                            0    ...      
2                 2                            0    ...      
3                 3                            1    ...      
4                 3                            0    ...      

   drop_location_type_G  drop_location_type_H  drop_location_type_I  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   drop_location_type_J  drop_location_type_K  drop_location_type_L  \
0                     0                     0                     0   
1                     0                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   

   drop_location_type_M  drop_location_type_N  sex_Female  sex_Male  
0                     0                     0           0         1  
1                     0                     0           0         1  
2                     0                     0           0         1  
3                     0                     0           1         0  
4                     0                     0           1         0  

[5 rows x 28 columns]

In [57]:
train_x=train.loc[:, train.columns != 'pricing_category']
train_y= train['pricing_category']

In [58]:
train_x.head()
train_x= train_x.fillna(train_x.median())

# Test data

In [59]:
test_data= pd.read_csv('test.csv')

In [60]:
test_id= test_data['id']
#test_data['diff']= test_data['anon_var_3']-test_data['anon_var_2']

In [61]:
# We don't need id as well
test= test_data.drop(['anon_var_1','id'],axis=1)


In [62]:
# Lets see the taxi type count
from collections import Counter
taxi_type_counter= Counter(test['taxi_type'])
customer_score_confidence_counter= Counter(test['customer_score_confidence'])
drop_location_type_counter= Counter(test['drop_location_type'])
print(taxi_type_counter)
print(customer_score_confidence_counter)
print(drop_location_type_counter)


Counter({'B': 12492, 'C': 11295, 'A': 8558, nan: 8058, 'D': 7643, 'E': 4619})
Counter({'B': 16207, 'C': 14328, 'A': 13978, nan: 8152})
Counter({'A': 30981, 'B': 11879, 'C': 2950, 'D': 2618, 'E': 1101, 'F': 805, 'G': 603, 'H': 493, 'I': 326, 'J': 292, 'L': 272, 'K': 264, 'M': 43, 'N': 38})


In [63]:
test_numerical= test.replace({'taxi_type': {'A':1, 'B': 2,'C':3,'D':4,'E':5}})

'''
test_numerical.replace({'customer_score_confidence':{'A':1, 'B': 3,'C':2}}, inplace=True)


test_numerical.replace({'drop_location_type':{'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E':5,
                                               'F': 6, 'G': 7, 'H': 8, 'I': 9, 'K': 10, 'J': 11, 
                                               'L': 12, 'N': 13, 'M': 14}},inplace=True)


test_numerical.replace({'drop_location_type':{'A': 11, 'B': 13, 'C': 12, 'D': 14, 'E':10,
                                               'F': 9, 'G': 8, 'H': 7, 'I': 6, 'K': 5, 'J': 4, 
                                               'L': 3, 'N': 2, 'M': 1}},inplace=True)
'''

"\ntest_numerical.replace({'customer_score_confidence':{'A':1, 'B': 3,'C':2}}, inplace=True)\n\n\ntest_numerical.replace({'drop_location_type':{'A': 1, 'B': 2, 'C': 3, 'D': 4, 'E':5,\n                                               'F': 6, 'G': 7, 'H': 8, 'I': 9, 'K': 10, 'J': 11, \n                                               'L': 12, 'N': 13, 'M': 14}},inplace=True)\n\n\ntest_numerical.replace({'drop_location_type':{'A': 11, 'B': 13, 'C': 12, 'D': 14, 'E':10,\n                                               'F': 9, 'G': 8, 'H': 7, 'I': 6, 'K': 5, 'J': 4, \n                                               'L': 3, 'N': 2, 'M': 1}},inplace=True)\n"

In [64]:
test_hot_encoded= pd.get_dummies(test_numerical)

In [65]:
#counting no of nans in final test array
np.count_nonzero(np.isnan(test_hot_encoded))
test_hot_encoded.isnull().sum()

distance                          0
taxi_type                      8058
months_of_activity             2387
customer_score                 8152
ratings_given_by_cust             0
num_of_cancelled_trips            0
anon_var_2                        0
anon_var_3                        0
customer_score_confidence_A       0
customer_score_confidence_B       0
customer_score_confidence_C       0
drop_location_type_A              0
drop_location_type_B              0
drop_location_type_C              0
drop_location_type_D              0
drop_location_type_E              0
drop_location_type_F              0
drop_location_type_G              0
drop_location_type_H              0
drop_location_type_I              0
drop_location_type_J              0
drop_location_type_K              0
drop_location_type_L              0
drop_location_type_M              0
drop_location_type_N              0
sex_Female                        0
sex_Male                          0
dtype: int64

In [67]:
test_x=test_hot_encoded

In [68]:
test_x.head()
test_x= test_x.fillna(test_x.median())

# Machine Learning

In [69]:
from scipy.sparse import csr_matrix
import numpy as np
import xgboost as xgb
from xgboost.sklearn import XGBClassifier
# Converting to sparse data and running xgboost

#X_csr = csr_matrix(train_x)
#xgb1 = XGBClassifier()
#xgb1.fit(xgtrain, y, eval_metric='auc')

model_xgb = xgb.XGBClassifier(n_estimators=2500, learning_rate=0.01, max_depth=7)

xgboost1 = model_xgb.fit(train_x, train_y)

from sklearn.metrics import accuracy_score

#X_test_csr= csr_matrix(test_x)

predxgboost_final = xgboost1.predict(test_x)

predxgboost_final= predxgboost_final.astype(float)

submission = pd.DataFrame(
   {'id': test_id, 'pricing_category': predxgboost_final},
   columns = ['id', 'pricing_category'])
submission.to_csv('last_try_2.csv', index = False)

In [ ]:
#data_hot_encoded= pd.get_dummies(small_data)